hydroflow_lang/src/graph/flat_graph.rs

```diff
@@ -96,178 +112,205 @@ impl FlatGraph {
+                // if let (
+                //     Some((src_port, GraphDet::Determined(src))),
+                //     Some((dst_port, GraphDet::Determined(dst))),
+                // ) = (lhs_ends.out, rhs_ends.inn)
+                // {
+                //     self.connect_operators(src_port, src, dst_port, dst);
+                // }
+
+                // Outer (first and last) ends.
+                let outer_ends = Ends {
                     inn: lhs_ends.inn,
                     out: rhs_ends.out,
-                }
+                };
+                // Inner (link) ends.
+                let link_ends = Ends {
+                    out: lhs_ends.out,
+                    inn: rhs_ends.inn,
+                };
+                self.links.push(link_ends);
+                outer_ends
             }
             Pipeline::Operator(operator) => {
                 let op_span = operator.span();
-                let key = self.nodes.insert(Node::Operator(operator));
+                let key = self.flat_graph.nodes.insert(Node::Operator(operator));
                 if let Some(current_varname) = current_varname {
-                    self.node_varnames.insert(key, current_varname.clone());
+                    self.flat_graph
+                        .node_varnames
+                        .insert(key, current_varname.clone());
                 }
                 Ends {
-                    inn: Some((PortIndexValue::Elided(op_span), key)),
-                    out: Some((PortIndexValue::Elided(op_span), key)),
+                    inn: Some((PortIndexValue::Elided(op_span), GraphDet::Determined(key))),
+                    out: Some((PortIndexValue::Elided(op_span), GraphDet::Determined(key))),
                 }
             }
         }
     }
 
+    /// Connects operator links as a final building step. Processes all the links stored in
+    /// `self.links` and actually puts them into the graph.
+    fn connect_operator_links(&mut self) {
+        for Ends { out, inn } in std::mem::take(&mut self.links) {
+            let out_opt = self.helper_resolve_name(out, false);
+            let inn_opt = self.helper_resolve_name(inn, true);
+            // `None` already have errors in `self.diagnostics`.
+            if let (Some((out_port, out_node)), Some((inn_port, inn_node))) = (out_opt, inn_opt) {
+                self.connect_operators(out_port, out_node, inn_port, inn_node);
+            }
+        }
+    }
+    /// Recursively resolve a variable name. For handling forward (and backward) name references
+    /// after all names have been assigned.
+    fn helper_resolve_name(
+        &mut self,
+        mut port_det: Option<(PortIndexValue, GraphDet)>,
+        is_in: bool,
+    ) -> Option<(PortIndexValue, GraphNodeId)> {
+        const DEFAULT_RECURSION_LIMIT: usize = 128;
+        // TODO(mingwei): Make recursion limit configurable if needed.
+
+        let mut span = Span::call_site();
+        for _ in 0..DEFAULT_RECURSION_LIMIT {
```

Comment:
Hmm is there a way we could replace this with logic to detect whether fixpoint has been reached on `varname_ends` without having all ports determined? Similarly, I wonder if there's a more precise definition of when a pipeline cannot be resoled (is only when a pipeline is used as both input and output, or are there other cases?).

Reply:
Yes its actually simpler than that, just cycle detection really. Also would save work to memoize I think?

---

I didn't [think too much](https://github.com/hydro-project/hydroflow/issues/158#issuecomment-1341848510) about what makes sense for cyclical/self-referential names, I just wrote the code and now some things work and some don't. The cases make sense if you think about them, but not at a glance

Specifically unresolved variables fail to resolve when those variables reference themself, creating a cycle. But keep in mind that each variable name is really two variables, one for the input and one for the output. When you write `my_var -> ...` you are only referencing the output side, and `... -> my_var` only references the input side. So an error will occur if there is a cycle such that the input end points at the input end itself, or the output end points out the output end itself.

`a = a -> identity()` works because the `a ->` references the output AKA last thing in `a`'s chain which is just `identity()`.

You do get an error if you have something in the following forms
* `a = a;` (and `a` is used elsewhere)
* `a = a -> a;` -- `a ->` is the output and the output of `a = ...` is `a` -- cycle! Likewise for input
* `a = a -> ... -> a;` -- same as above

This can be made more complicated once you introduce other variables, mutual recursion style...
```rust
b = a -> map(|x: usize| x) -> a;
b = a;
```
---

There are definitely some dubious cases that work perfectly fine...
```rust
self_ref = map(|a: usize| a) -> [0]self_ref[1] -> map(|b: usize| b);
```
The input and output of `self_ref` are perfectly defined as those two maps, so everything resolves.
```mermaid
%%{init:{'theme':'base','themeVariables':{'clusterBkg':'#ddd','clusterBorder':'#888'}}}%%
flowchart TD
classDef pullClass fill:#02f,color:#fff,stroke:#000
classDef pushClass fill:#ff0,stroke:#000
linkStyle default stroke:#aaa,stroke-width:4px,color:red,font-size:1.5em;
subgraph sg_1v1 ["sg_1v1 stratum 0"]
    1v1["(1v1) <tt>map(| a : usize | a)</tt>"]:::otherClass
    4v1["(4v1) <tt>handoff</tt>"]:::otherClass
    4v1--0--->1v1
    1v1--->4v1
    subgraph sg_1v1_var_self_ref ["var <tt>self_ref</tt>"]
        1v1
    end
end
subgraph sg_2v1 ["sg_2v1 stratum 0"]
    2v1["(2v1) <tt>map(| b : usize | b)</tt>"]:::otherClass
    3v1["(3v1) <tt>handoff</tt>"]:::otherClass
    3v1--->2v1
    2v1--1--->3v1
    subgraph sg_2v1_var_self_ref ["var <tt>self_ref</tt>"]
        2v1
    end
end
```
Its justifiable logically, but silly
![image](https://user-images.githubusercontent.com/6778341/219899176-7364ecfa-a090-4329-a8e6-c5804f9917a0.png)